File: Code-Eval/moderate/overlapping_rectangles.py
# Python 3
import sys

class Rectangle:
    def __init__(self, x1, y1, x2, y2):
        self.x1 = x1
        self.y1 = y1
        self.x2 = x2
        self.y2 = y2
        
    def pointWithin(self, x, y):
        if x >= self.x1 and x <= self.x2:
            if y >= self.y2 and y <= self.y1:
                return True
        return False

test_cases = open(sys.argv[1], 'r')
for test in test_cases:
    coord = test.strip().split(',')
    
    for i in range(len(coord)):
        coord[i] = int(coord[i])
        
    minXVal = min(coord[0], coord[2], coord[4], coord[6])
    minYVal = min(coord[1], coord[3], coord[5], coord[7])
    #print("Minimum x and y: ", minXVal, minYVal)
    
    if minXVal < 0:
        minXVal = minXVal * -1
    if minYVal < 0:
        minYVal = minYVal * -1
        
    coord[0] = coord[0] + minXVal
    coord[2] = coord[2] + minXVal
    coord[4] = coord[4] + minXVal
    coord[6] = coord[6] + minXVal
    
    coord[1] = coord[1] + minYVal
    coord[3] = coord[3] + minYVal
    coord[5] = coord[5] + minYVal
    coord[7] = coord[7] + minYVal
    
    #for i in coord:
    #    print(i)
    
    r1 = Rectangle(coord[0], coord[1], coord[2], coord[3])
    r2 = Rectangle(coord[4], coord[5], coord[6], coord[7])
    print(r1.pointWithin(r2.x1, r2.y1) or r1.pointWithin(r2.x2, r2.y2) or r1.pointWithin(r2.x2, r2.y1) or r1.pointWithin(r2.x1, r2.y2))

test_cases.close()

File: Code-Eval/moderate/overlapping_rectangles.cpp
#include <cstdio>
#include <iostream>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>

struct rectangle {
    int x1;
    int y1;
    int x2;
    int y2;
};

std::string checkOverlapping(rectangle r1, rectangle r2);

int main(int argc, char *argv[])
{
    std::ifstream file;
    std::string lineBuffer;

    file.open(argv[1]);

    while (!file.eof()) {

        getline(file, lineBuffer);

        if (lineBuffer == "") {
            continue;
        } else {
            // ready line for parsing
            for (int index = 0; index < lineBuffer.length(); index++) {
                //std::cout << lineBuffer[index];
                if (lineBuffer[index] == ',') {
                    lineBuffer[index] = ' ';
                }
            }
            //std::cout << std::endl;

            rectangle rect1;
            rectangle rect2;
            std::string number;
            std::stringstream ss(lineBuffer);

            // parse data into data structure
            ss >> rect1.x1 >> rect1.y1 >> rect1.x2 >> rect1.y2;
            ss >> rect2.x1 >> rect2.y1 >> rect2.x2 >> rect2.y2;

            //std::cout << "Rect 1: " << rect1.x1 << "," << rect1.y1 << " " << rect1.x2 << "," << rect1.y2 << std::endl;
            //std::cout << "Rect 2: " << rect2.x1 << "," << rect2.y1 << " " << rect2.x2 << "," << rect2.y2 << std::endl;

            int minXVal, minYVal;

            minXVal = rect1.x1;
            minXVal = rect1.x2 < minXVal ? rect1.x2 : minXVal;
            minXVal = rect2.x1 < minXVal ? rect2.x1 : minXVal;
            minXVal = rect2.x2 < minXVal ? rect2.x2 : minXVal;

            minYVal = rect1.y1;
            minYVal = rect1.y2 < minYVal ? rect1.y2 : minYVal;
            minYVal = rect2.y1 < minYVal ? rect2.y1 : minYVal;
            minYVal = rect2.y2 < minYVal ? rect2.y2 : minYVal;

            minXVal = minXVal < 0 ? minXVal * -1 : minXVal;
            minYVal = minYVal < 0 ? minYVal * -1 : minYVal;

            //std::cout << "Min vals: " << minXVal << " " << minYVal << std::endl;

            rect1.x1 += minXVal;
            rect1.y1 += minYVal;
            rect1.x2 += minXVal;
            rect1.y2 += minYVal;

            rect2.x1 += minXVal;
            rect2.y1 += minYVal;
            rect2.x2 += minXVal;
            rect2.y2 += minYVal;

            // check data and print result
            std::cout << checkOverlapping(rect1, rect2) << std::endl;
        }
        //if (!file.eof()) {
        //    std::cout << std::endl;
        //}
    }

    return 0;
}

std::string checkOverlapping(rectangle r1, rectangle r2)
{
    if (r1.x1 > r2.x1) {
        rectangle temp = r1;
        r1 = r2;
        r2 = temp;
    }

    if ((r2.y1 < r1.y2) || /* r2 is below r1 */
        (r2.x1 > r1.x2) || /* r2 is to the right of r1 */
        (r2.y2 > r1.y1)) { /* r2 is above and right of r1 */
        return "False";
    } else {
        return "True";
    }
}
